Assignment3/PointAndGrid.py

```diff
@@ -0,0 +1,240 @@
+            for j in range(numCols):
+                if self.grid[i][j] == element:
+                    result.append(Point(self, (i, j), element))
+        return result
+
+    def findNeighborsOfPoint(self, tuple):
+        """
+        Return list of Points that are neighbours of element with the given coordinates.
+        :param tuple: Tuple
+        :return: list(Point)
+        """
+        row = tuple[0]
+        col = tuple[1]
+        result = []
+        possible_coord = [(row-1, col-1),(row-1, col),(row, col-1),(row-1, col+1),
+                          (row+1, col-1),(row+1, col),(row, col+1),(row+1, col+1)]
+        
+        for coord in possible_coord:
+            if self._validCoordinates(coord):
+                result.append(Point(self, coord, self.getElement(coord)))
+        return result if result else None
+    
+    def ifWordInGrid(self, word):
+        """
+        Return True if the word can be formed in grid, False otherwise.
+        :param word: str
+        :return: bool
+        """
+        return self.findString(word) is None
+    
+    ############################## THE SOLUTION TO ASSIGNMENT 3 ##########################################
+
+    def findString(self, word):
+        """
+        Find the word in the grid where we can starting from one position move to any neighbors,
+        bub can't visit the same cell twice. Return the path if exists, None otherwise.
+        :param word: string
+        :return: bool
+        """
+        startPoints = self.findElementInGrid(word[0])
+        path = []
+        for startP in startPoints: 
+            # we are here only if the first letter of word is in grid
+            path = [startP]
+            path = self._findPathRecursively(startP, word, path, [])
+        if self._pathIsWord(path, word):
+            return path
+
+    def _pathIsWord(self, path, word):
+        return [p.value for p in path] == list(word)
+
+    def _findPathRecursively(self, point, word, path, visited):
+        """
+        Helper-function to find a word in a grid. Finds the longest path to form the word (may be only a part)
+        :param point: Point
+        :param word: str
+        :param path: list(Point)
+        :param visited: list(Point)
+        :return: list(Point) or None
+        """
+        if self._pathIsWord(path, word):
+            return path
+
+        possible_ways = self.findPossibleWays(path, point, word, visited)
+        if not possible_ways:
+            # we at the same time pop the last element from path and append it to list visited
+            visited.append(path.pop(-1))
+            if path:
+                self._findPathRecursively(path[-1], word, path, visited)
+
+        for neighbor in possible_ways:
+            if neighbor not in path and neighbor not in visited:
+                path.append(neighbor)
+                self._findPathRecursively(neighbor, word, path, visited)
+                
+        return path if path else None

```

Comment:
I still can't see how this recursive function could work correctly. If `path` was not `None` when it was passed in then it will still be not `None` and will be passed back as it was passed in to the first stack frame.

I think as it stands, the method will only work correctly if looking for 1-character words.